docs/spec/fastsync/fastsync.md

```diff
@@ -0,0 +1,1164 @@
+> the notions *validator pair* and *validator node*, respectively, to
+> distinguish these notions in the cases where they are not clear from
+> the context.
+
+
+#### **[TMBC-VALIDATOR-PAIR]**:
+
+Given a full node, a 
+*validator pair* is a pair *(address, voting_power)*, where 
+  - *address* is the address (public key) of a full node, 
+  - *voting_power* is an integer (representing the full node's
+  voting power in a certain consensus instance).
+  
+> In the Golang implementation the data type for *validator
+> pair* is called `Validator`
+
+#### **[TMBC-VALIDATOR-SET]**:
+
+A *validator set* is a set of validator pairs. For a validator set
+*vs*, we write *TotalVotingPower(vs)* for the sum of the voting powers
+of its validator pairs.
+
+#### **[TMBC-CORRECT]**:
+We define a predicate *correctUntil(n, t)*, where *n* is a node and *t* is a 
+time point. 
+The predicate *correctUntil(n, t)* is true if and only if the node *n* 
+follows all the protocols (at least) until time *t*.
+
+#### **[TMBC-TIME-PARAMS]**:
+A blockchain has the following configuration parameters:
+ - *unbondingPeriod*: a time duration.
+ - *trustingPeriod*: a time duration smaller than *unbondingPeriod*.
+
+#### **[TMBC-FM-2THIRDS]**:
+If a block *h* is in the chain,
+then there exists a subset *CorrV*
+of *h.NextValidators*, such that:
+  - *TotalVotingPower(CorrV) > 2/3
+    TotalVotingPower(h.NextValidators)*; 
+  - For every validator pair *(n,p)* in *CorrV*, it holds *correctUntil(n,
+    h.Time + trustingPeriod)*.
+
+#### **[TMBC-CORR-FULL]**: 
+Every correct full node locally stores a prefix of the
+current list of headers from [**[TMBC-SEQ]**][TMBC-SEQ-link].
+
+
+
+
+
+## Informal Problem statement
+
+
+A full node has as input a block of the blockchain at height *h* and
+the corresponding application state (or the prefix of the current
+blockchain until height *h*). It has access to a set *peerIDs* of full
+nodes called *peers* that it knows of.  The full node uses the peers
+to read blocks of the Tendermint blockchain (in a safe way, that is,
+it checks the soundness conditions), until it has read the most recent
+block and then terminates.
+
+
+## Sequential Problem statement
+
+
+
+*Fastsync* gets as input a block of height *h* and the corresponding
+application state *s* that corresponds to the block and state of that
+height of the blockchain, and produces
+as output (i) a list *L* of blocks starting at height *h* to some height
+*terminationHeight*, and (ii) the application state when applying the
+transactions of the list *L* to *s*. Fastsync has to satisfy the following
+properties [FS-SEQ-?]:
+
+
+#### **[FS-SEQ-LIVE]**: 
+*Fastsync* eventually terminates.
+
+ 
+#### **[FS-SEQ-TERM]**:
+Let *bh* be the height of the blockchain at the time *Fastsync*
+starts. By assumption we have *bh >= h*.
+When *Fastsync* terminates, it outputs a list of all blocks from
+height *h* to some height *terminationHeight >= bh - 1*.
```

Comment:
I don't get this property. Not sure why you need to use both bh and h as starting height.